core/unit_test/TestHostSharedPtrAccessOnDevice.hpp

```diff
@@ -0,0 +1,154 @@
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <impl/Kokkos_HostSharedPtr.hpp>
+#include <impl/Kokkos_SharedAlloc.hpp>
+#include <Kokkos_Core.hpp>
+
+#include <gtest/gtest.h>
+
+using Kokkos::Impl::HostSharedPtr;
+
+namespace {
+
+class Data {
+  Kokkos::Array<char, 64> d;
+
+ public:
+  KOKKOS_FUNCTION void write(char const* c) {
+    for (int i = 0; i < 64 && c; ++i, ++c) {
+      d[i] = *c;
+    }
+  }
+};
+
+template <class SmartPtr>
+struct CheckAccessStoredPointerAndDereferenceOnDevice {
+  SmartPtr m_device_ptr;
+  using ElementType = typename SmartPtr::element_type;
+  static_assert(std::is_same<ElementType, Data>::value, "");
+
+  CheckAccessStoredPointerAndDereferenceOnDevice(SmartPtr device_ptr)
+      : m_device_ptr(device_ptr) {
+    int errors;
+    Kokkos::parallel_reduce(Kokkos::RangePolicy<TEST_EXECSPACE>(0, 1), *this,
+                            errors);
+    EXPECT_EQ(errors, 0);
+  }
+
+  KOKKOS_FUNCTION void operator()(int, int& e) const {
+    auto raw_ptr = m_device_ptr.get();  // get
+
+    auto tmp = new (raw_ptr) ElementType();
+
+    auto& obj = *m_device_ptr;  // operator*
+    if (&obj != raw_ptr) ++e;
+
+    m_device_ptr->write("hello world");  // operator->
+
+    tmp->~ElementType();
+  }
+};
+
+template <class Ptr>
+CheckAccessStoredPointerAndDereferenceOnDevice<Ptr>
+check_access_stored_pointer_and_dereference_on_device(Ptr p) {
+  return {p};
+}
+
+template <class SmartPtr>
+struct CheckCopyOnDevice {
+  SmartPtr m_device_ptr;
+  using ElementType = typename SmartPtr::element_type;
+  static_assert(std::is_same<ElementType, Data>::value, "");
+
+  CheckCopyOnDevice(SmartPtr device_ptr) : m_device_ptr(device_ptr) {
+    int errors;
+    Kokkos::parallel_reduce(Kokkos::RangePolicy<TEST_EXECSPACE>(0, 1), *this,
+                            errors);
+    EXPECT_EQ(errors, 0);
+  }
+
+  KOKKOS_FUNCTION void operator()(int, int&) const {
+    SmartPtr p1 = m_device_ptr;   // copy constructor
+    SmartPtr p2 = std::move(p1);  // move constructor
+
+    p1 = p2;             // copy assignment
+    p2 = std::move(p1);  // move assignment
```

Comment:
Minor/non-blocking: if `TEST_EXECSPACE` is a host space, should we check use counts, etc., and increment the error count if they have unexpected values?

Reply:
I'd rather not